### PR TITLE
Add rewrite rule to optimize a common case of Literal array contains

### PR DESCRIPTION
**Pull request by tpoterba**

### CODE REVIEW

File: hail/src/main/scala/is/hail/expr/ir/Simplify.scala [hunk 1]
```diff
@@ -167,6 +167,15 @@ object Simplify {
...
       ArrayRef(a, i)
 
+    case ToArray(x) if x.typ.isInstanceOf[TArray] => x
+
+    case ApplyIR("contains", Seq(ToArray(x), element)) if x.typ.isInstanceOf[TSet] => invoke("contains", x, element)
+
+    case ApplyIR("contains", Seq(Literal(t, v), element)) if t.isInstanceOf[TArray] =>
+      invoke("contains", Literal(TSet(t.asInstanceOf[TArray].elementType, t.required), v.asInstanceOf[IndexedSeq[_]].toSet), element)
+
+    case ApplyIR("contains", Seq(ToSet(x), element)) if x.typ.isInstanceOf[TArray] => invoke("contains", x, element)
```
Comment (jigold) [inline hail/src/main/scala/is/hail/expr/ir/Simplify.scala]: What's the rationale for this change? It seems to me it would be slower.

File: hail/src/main/scala/is/hail/expr/ir/functions/ArrayFunctions.scala [hunk 2]
```diff
@@ -73,6 +88,8 @@ object ArrayFunctions extends RegistryFunctions {
       extend(a, MakeArray(Seq(c), TArray(c.typ)))
     }
 
+    registerIR("contains", TArray(tv("T")), tv("T"), TBoolean()) { (a, e) => contains(a, e)}
```
Comment (jigold) [inline hail/src/main/scala/is/hail/expr/ir/functions/ArrayFunctions.scala]: formatting at end of line. missing space

File: hail/src/test/scala/is/hail/expr/ir/ArrayFunctionsSuite.scala [hunk 3]
```diff
@@ -253,6 +253,45 @@ class ArrayFunctionsSuite extends TestNGSuite {
...
+      invoke("contains", In(0, t), Str("a")),
+      args = FastIndexedSeq(FastIndexedSeq() -> t),
+      expected=false)
+
+    assertEvalsTo(
+      invoke("contains", In(0, t), Str("a")),
+      args = FastIndexedSeq(FastIndexedSeq(null) -> t),
+      expected=false)
+
+
```
Comment (jigold) [inline hail/src/test/scala/is/hail/expr/ir/ArrayFunctionsSuite.scala]: delete extra new line.

File: hail/src/test/scala/is/hail/expr/ir/ArrayFunctionsSuite.scala [hunk 4]
```diff
@@ -253,6 +253,45 @@ class ArrayFunctionsSuite extends TestNGSuite {
...
   def flatten(in: IndexedSeq[IndexedSeq[Integer]], expected: IndexedSeq[Int]) {
     assertEvalsTo(invoke("flatten", MakeArray(in.map(toIRArray(_)), TArray(TArray(TInt32())))), expected)
+  }
+
+  @Test def testContains() {
+    val t = TArray(TString())
+
+    assertEvalsTo(
+      invoke("contains", In(0, t), Str("a")),
+      args = FastIndexedSeq(FastIndexedSeq() -> t),
```
Comment (jigold) [inline hail/src/test/scala/is/hail/expr/ir/ArrayFunctionsSuite.scala]: This only tests the new contains for arrays in IR. Do we need tests of the simplify methods?

File: hail/src/main/scala/is/hail/expr/ir/Simplify.scala (hunk 1 above)
Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/Simplify.scala]: Array contains requires N comparisons, while `ToSet(array) contains` will do N log N to create the set (with a full copy!), and then another logN to binary search for the element.

If someone is going to create a set to use with `.contains` and then immediately discard it, we can save that work!

File: hail/src/test/scala/is/hail/expr/ir/ArrayFunctionsSuite.scala (hunk 4 above)
Comment (tpoterba) [reply hail/src/test/scala/is/hail/expr/ir/ArrayFunctionsSuite.scala]: I'll add something to SimplifySuite, but in general we don't have a good story for isolating a specific rewrite behavior in testing. Running `assertEvalsTo` will compare the result of evaluating an IR without optimization with full optimization, and it's totally possible that other rules will eliminate the specific rewrite we might be interested in testing.

File: hail/src/main/scala/is/hail/expr/ir/Simplify.scala (hunk 1 above)
Comment (jigold) [reply hail/src/main/scala/is/hail/expr/ir/Simplify.scala]: ok. that's what i figured but wanted to make sure.
